tests/nextflowtower/test_utils.py

```diff
@@ -0,0 +1,191 @@
+        result = tower_utils.get_workflow(workflow_id)
+        assert result == self.EG_WORKFLOW
+
+    EG_DICT = {
+        "foo": [1, 2, 3],
+        "bar": None,
+        "baz": {"tic": 1.1, "tac": 2.2, "toe": 3.3},
+    }
+
+    def test_update_dict_none(self, tower_utils):
+        # Expecting no change due to None value
+        overrides = {"foo": None}
+        result = tower_utils.update_dict(self.EG_DICT, overrides)
+        assert result == self.EG_DICT
+
+    def test_update_dict_invalid(self, tower_utils):
+        # Expecting no change due to None value
+        overrides = {"oof": "gah"}
+        with pytest.raises(ValueError):
+            tower_utils.update_dict(self.EG_DICT, overrides)
+
+    def test_update_dict_toplevel(self, tower_utils):
+        # Expecting top-level changes
+        overrides = {"foo": [4, 5, 6], "bar": False}
+        result = tower_utils.update_dict(self.EG_DICT, overrides)
+        assert result["foo"] == overrides["foo"]
+        assert result["bar"] == overrides["bar"]
+
+    def test_update_dict_nested(self, tower_utils):
+        # Expecting nested changes
+        overrides = {"baz": {"tic": 7.7, "tac": {"some": "thing"}}}
+        result = tower_utils.update_dict(self.EG_DICT, overrides)
+        assert result["baz"]["tic"] == overrides["baz"]["tic"]
+        assert result["baz"]["tac"] == overrides["baz"]["tac"]
+
+    def test_update_dict_copy(self, tower_utils):
+        # Expecting input dictionary to not change
+        eg_dict_copy = deepcopy(self.EG_DICT)
+        overrides = {"bar": "random"}
+        result = tower_utils.update_dict(self.EG_DICT, overrides)
+        assert result["bar"] == overrides["bar"]
+        assert self.EG_DICT == eg_dict_copy
+
+    def test_init_launch_workflow_data(self, mocker, tower_utils):
+        compute_env = self.EG_COMPUTE_ENV["computeEnv"]
+        compute_env_id = compute_env["id"]
+        ce_config = compute_env["config"]
+        mocked = mocker.patch.object(tower_utils, "get_compute_env")
+        mocked.return_value = compute_env
+        result = tower_utils.init_launch_workflow_data(compute_env_id)
+        launch = result["launch"]
+        assert launch["postRunScript"] == ce_config["postRunScript"]
+        assert launch["preRunScript"] == ce_config["preRunScript"]
+        assert launch["workDir"] == ce_config["workDir"]
+        assert launch["dateCreated"] is not None
+
+    EG_LAUNCH = {"workflowId": "7g2R5Z1J"}
+
+    def test_launch_workflow(self, mocker, tower_utils):
+        # Mock TowerUtils class but use static reference for `launch_workflow()`
+        # Credit: https://stackoverflow.com/a/35782720
+        compute_env_id = self.EG_COMPUTE_ENV["computeEnv"]["id"]
+        pipeline = self.EG_WORKFLOW["workflow"]["projectName"]
+        mocked_utils = mocker.MagicMock(tower_utils, autospec=True)
+        # Test the function call
+        TowerUtils.launch_workflow(mocked_utils, compute_env_id, pipeline)
```

Comment:
I feel like this test could be better (_e.g._ like checking for values), but the function being tested is mostly calls to other methods, so almost everything is a Python Mock.

Reply:
Nit: You can use `assert_called_once_with` to be more explicit to make sure all the parameters from one function to another are called, but I think it's technically fine.